Resolve initial rename value

Krzysztof-Cieslak opened the pull request:
PR implements the functionality discussed in #7340. It's just a proposal, I've done it mostly to learn a bit about architecture around extensions API, so feel free to close it if it's not going in good direction.

CC: @aeschli & @jrieken

jrieken commented:
Looks pretty good after a quick read... I have to check with my schedule but we make this happen in November...

jrieken commented on src/vs/editor/contrib/rename/browser/rename.ts:
```diff
@@ -79,6 +79,33 @@ export function rename(model: IReadOnlyModel, position: Position, newName: strin
 }
 
+function resolveInitialRenameValue(model: IReadOnlyModel, position: Position): TPromise<RenameInitialValue> {
+	const supports = RenameProviderRegistry.ordered(model);
```
We should just go for the first rename provider so that we always use the same for resolving the location and for the actual rename.

jrieken commented on src/vs/editor/contrib/rename/browser/rename.ts:
```diff
@@ -117,34 +144,64 @@ class RenameController implements IEditorContribution {
+
+		let initialValue = await resolveInitialRenameValue(this.editor.getModel(), this.editor.getPosition());
+
+		if(initialValue) {
```
We can maybe move this to a 'default' implementation of resolveRenameValue

Krzysztof-Cieslak replied:
I'm not really sure, the code below is just setting values local to the `run` function. It's only small extension of the logic that was already there before.

jrieken commented on src/vs/vscode.d.ts:
```diff
@@ -2482,6 +2482,11 @@ declare module 'vscode' {
 		appendVariable(name: string, defaultValue: string | ((snippet: SnippetString) => any)): SnippetString;
 	}
 
+	export interface RenameInitialValue {
```
Needs maybe some better naming here but idea is good

jrieken replied:
I am actually wondering when we would return both, so range and text (which then would be different from the text in range?). Do you have a use-case in mind?

jrieken commented on src/vs/workbench/api/node/extHost.protocol.ts:
```diff
@@ -594,6 +594,7 @@ export interface ExtHostLanguageFeaturesShape {
 	$resolveWorkspaceSymbol(handle: number, symbol: modes.SymbolInformation): TPromise<IWorkspaceSymbol>;
 	$releaseWorkspaceSymbols(handle: number, id: number): void;
 	$provideRenameEdits(handle: number, resource: URI, position: IPosition, newName: string): TPromise<modes.WorkspaceEdit>;
+	$resolveInitialRenameValue(handle: number, resource: URI, position: IPosition): TPromise<modes.RenameInitialValue>;
```
We removed some automagic marshalling and `UriComponent` must be used now

jrieken commented:
@Krzysztof-Cieslak Are you still with us on this one? We have scheduled this for January.

Krzysztof-Cieslak commented:
@jrieken, yes, I’ve been on vacations after Xmas. Give me day or two and I’ll get back to it.

Krzysztof-Cieslak commented:
@jrieken I've rebased PR to latest master and moved to using the first provider for resolving. I'm not sure if I agree/understand second suggestion.

Also, happy to hear any suggestions for the naming :)

jrieken commented on src/vs/workbench/api/node/extHostLanguageFeatures.ts:
```diff
@@ -1003,14 +1023,18 @@ export class ExtHostLanguageFeatures implements ExtHostLanguageFeaturesShape {
 		return this._withAdapter(handle, RenameAdapter, adapter => adapter.provideRenameEdits(URI.revive(resource), position, newName));
 	}
 
+	$resolveInitialRenameValue(handle: number, resource: URI, position: IPosition): TPromise<modes.RenameInitialValue> {
```
The `URI` argument won't be of that type at runtime. That's because the function will be called by our JSON-rpc internals and the runtime type will be `UriComponents` (from which you then create an `URI` using `revive`).

jrieken commented on src/vs/vscode.d.ts:
```diff
@@ -2567,6 +2572,8 @@ declare module 'vscode' {
 		 */
 		provideRenameEdits(document: TextDocument, position: Position, newName: string, token: CancellationToken): ProviderResult<WorkspaceEdit>;
+
+		resolveInitialRenameValue?(document: TextDocument, position: Position, token: CancellationToken): ProviderResult<RenameInitialValue>;
```
Maybe about `RenameInformation`? That would make it: 
```ts
resolveRenameInformation?(document: TextDocument, position: Position, token: CancellationToken): ProviderResult<RenameInformation>;

provideRenameInformation?(document: TextDocument, position: Position, token: CancellationToken): ProviderResult<RenameInformation>;

```

jrieken replied:
Or something more clever sounding, like `RenameForesight`, `RenamePreflight`, etc...

jrieken commented:
Pulling this in but marking the new API as proposed which buys as some time to polish names etc.  @Krzysztof-Cieslak Thanks for good work.

Krzysztof-Cieslak commented:
@jrieken, thanks for merging!
